BREAKING: `WSManListener`: Convert to a Class Resource

dan-hughes opened the pull request:
<!--
    Thanks for submitting a Pull Request (PR) to this project.
    Your contribution to this project is greatly appreciated!

    Please prefix the PR title with the resource name,
    e.g. 'ResourceName: My short description'.
    If this is a breaking change, then also prefix the PR title
    with 'BREAKING CHANGE:',
    e.g. 'BREAKING CHANGE: ResourceName: My short description'.

    You may remove this comment block, and the other comment blocks, but please
    keep the headers and the task list.
-->

#### Pull Request (PR) description

<!--
    Replace this comment block with a description of your PR.
    Also, make sure you have updated the CHANGELOG.md, see the
    task list below. An entry in the CHANGELOG.md is mandatory
    for all PRs.
-->
Convert `WSManListener` to a class resource
Rename parameter `DN` to `BaseDN`

#### This Pull Request (PR) fixes the following issues

<!--
    If this PR does not fix an open issue, replace this comment block with None.
    If this PR resolves one or more open issues, replace this comment block with
    a list of the issues using a GitHub closing keyword, e.g.:

- Fixes #123
- Fixes #124
-->

Contributes to #98 
Fixes #89

#### Task list

<!--
    To aid community reviewers in reviewing and merging your PR, please take
    the time to run through the below checklist and make sure your PR has
    everything updated as required.

    Change to [x] for each task in the task list that applies to your PR.
    For those task that don't apply to you PR, leave those as is.
-->

- [x] Added an entry to the change log under the Unreleased section of the
      file CHANGELOG.md. Entry should say what was changed and how that
      affects users (if applicable), and reference the issue being resolved
      (if applicable).
- [x] Resource documentation added/updated in README.md.
- [x] Resource parameter descriptions added/updated in README.md, schema.mof
      and comment-based help.
- [x] Comment-based help added/updated.
- [x] Localization strings added/updated in all localization files as appropriate.
- [x] Examples appropriately added/updated.
- [x] Unit tests added/updated. See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [x] Integration tests added/updated (where possible). See [DSC Community Testing Guidelines](https://dsccommunity.org/guidelines/testing-guidelines).
- [x] New/changed code adheres to [DSC Community Style Guidelines](https://dsccommunity.org/styleguidelines).

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dsccommunity/WSManDsc/106)
<!-- Reviewable:end -->

dan-hughes commented:
@PlagueHO, finally ready to go.

dan-hughes commented:
@johlju, is this something you can also look at please?

johlju commented:
Will do, later in the week or weekend. 😊

johlju reviewed (changes requested):
> Reviewed 15 of 28 files at r1, 19 of 20 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/dsccommunity/WSManDsc/106)* status: all files reviewed, 4 unresolved discussions (waiting on @dan-hughes and @PlagueHO)

___
*[`source/Classes/020.WSManListener.ps1` line 232 at r3](https://reviewable.io/reviews/dsccommunity/WSManDsc/106#-OGOjm3Y5inzSvIBv3Qn:-OGOjm3Y5inzSvIBv3Qo:b-48vfpc) ([raw file](https://github.com/dsccommunity/WSManDsc/blob/514c668d3a651a2ab0c70d310b4acd6d9fe1c9dc/source/Classes/020.WSManListener.ps1#L232)):*
> ```PowerShell
>                 'Issuer'
>                 'BaseDN'
>             )
> ```

Shouldn't there also be `SubjectFormat` and `MatchAlternate` in this list, they should also not be used when thumbprint or hostname is specified?



_Code quote:_
```PowerShell
            MutuallyExclusiveList1 = @(
                'Issuer'
                'BaseDN'
            )
```
___
*[`source/Classes/020.WSManListener.ps1` line 249 at r3](https://reviewable.io/reviews/dsccommunity/WSManDsc/106#-OGOl1QU2UkHPr2gXvVG:-OGOl1QU2UkHPr2gXvVH:by6jqj2) ([raw file](https://github.com/dsccommunity/WSManDsc/blob/514c668d3a651a2ab0c70d310b4acd6d9fe1c9dc/source/Classes/020.WSManListener.ps1#L249)):*
> ```PowerShell
>             Transport = $this.Transport
>             Address   = $this.Address
>         }
> ```

In this method we need to validate all properties needed to create an instance actually has a value. We cannot do it in `AssertProperties()` as that would prevent as enforcing only a specific  property. :thinking: 



_Code quote:_
```PowerShell
        $selectorSet = @{
            Transport = $this.Transport
            Address   = $this.Address
        }
```
___
*[`source/Classes/020.WSManListener.ps1` line 269 at r3](https://reviewable.io/reviews/dsccommunity/WSManDsc/106#-OGOlpVA4B56ubsTwVVH:-OGOlpVA4B56ubsTwVVI:bg23dbq) ([raw file](https://github.com/dsccommunity/WSManDsc/blob/514c668d3a651a2ab0c70d310b4acd6d9fe1c9dc/source/Classes/020.WSManListener.ps1#L269)):*
> ```PowerShell
>                 if ([System.String]::IsNullOrEmpty($this.Hostname))
>                 {
>                     $valueSet.HostName = [System.Net.Dns]::GetHostByName($env:COMPUTERNAME).Hostname
> ```

Suggest to use `Get-ComputerName` here instead (from DscResource.Common)



_Code quote:_
```PowerShell
$env:COMPUTERNAME
```
___
*[`source/Classes/020.WSManListener.ps1` line 288 at r3](https://reviewable.io/reviews/dsccommunity/WSManDsc/106#-OGOmH9bBU8rXt1ozZVW:-OGOmH9bBU8rXt1ozZVX:b-scxx3i) ([raw file](https://github.com/dsccommunity/WSManDsc/blob/514c668d3a651a2ab0c70d310b4acd6d9fe1c9dc/source/Classes/020.WSManListener.ps1#L288)):*
> ```PowerShell
>     }
> 
>     hidden [void] RemoveInstance()
> ```

Do we need to validate here as well, that the required properties to remove an instance is actually set? :thinking: 



_Code quote:_
```PowerShell
RemoveInstance()
```
___
*[`source/Private/Find-Certificate.ps1` line 24 at r3](https://reviewable.io/reviews/dsccommunity/WSManDsc/106#-OGOeUqC-UnodX1NNem0:-OGOeUqD5zU9NhZ7mDzh:b-97nlpy) ([raw file](https://github.com/dsccommunity/WSManDsc/blob/514c668d3a651a2ab0c70d310b4acd6d9fe1c9dc/source/Private/Find-Certificate.ps1#L24)):*
> ```PowerShell
>         The Thumbprint of the certificate to use for the HTTPS WS-Man Listener.
> #>
> function Find-Certificate
> ```

We also have a `Find-Certificate` in [CertificateDsc](https://github.com/dsccommunity/CertificateDsc/blob/0dcd82e1d6fb67e1f0f1dd597c8cbe3b97abb163/source/Modules/CertificateDsc.Common/CertificateDsc.Common.psm1#L280C10-L281C2) - it might be possible to concat these inte one public function in DscResource.Common. :thinking:  But out of scope for this PR.



_Code quote:_
```PowerShell
Find-Certificate
```


<!-- Sent from Reviewable.io -->

johlju commented:
Justa few comments. Heads up, there might be some issues with PowerShell Gallery - i trying to release a version before merging this breaking change, but it fails with random errors (re-run 3 times, failing on different module name each time): https://dev.azure.com/dsccommunity/WsManDsc/_build/results?buildId=9921&view=logs&j=d7ffe41d-f206-59ed-99c3-e44ba5bb1f40&t=e2ebcab5-cb94-5995-9070-3fa60620c9c4

dan-hughes reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/dsccommunity/WSManDsc/106)* status: all files reviewed, 4 unresolved discussions (waiting on @johlju and @PlagueHO)

___
*[`source/Classes/020.WSManListener.ps1` line 232 at r3](https://reviewable.io/reviews/dsccommunity/WSManDsc/106#-OGOjm3Y5inzSvIBv3Qn:-OGPrZtj566Ntqwtj9EP:bt3bcw7) ([raw file](https://github.com/dsccommunity/WSManDsc/blob/514c668d3a651a2ab0c70d310b4acd6d9fe1c9dc/source/Classes/020.WSManListener.ps1#L232)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Shouldn't there also be `SubjectFormat` and `MatchAlternate` in this list, they should also not be used when thumbprint or hostname is specified?
</blockquote></details>

I believe you are correct. 
___
*[`source/Classes/020.WSManListener.ps1` line 249 at r3](https://reviewable.io/reviews/dsccommunity/WSManDsc/106#-OGOl1QU2UkHPr2gXvVG:-OGPp_YY2F-HTnFFgWk6:boac43e) ([raw file](https://github.com/dsccommunity/WSManDsc/blob/514c668d3a651a2ab0c70d310b4acd6d9fe1c9dc/source/Classes/020.WSManListener.ps1#L249)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

In this method we need to validate all properties needed to create an instance actually has a value. We cannot do it in `AssertProperties()` as that would prevent as enforcing only a specific  property. :thinking:
</blockquote></details>

I don't believe so, if the port is not supplied, then it is populated with `Get-DefaultPort` in `Get().GetCurrentState()`. 
___
*[`source/Classes/020.WSManListener.ps1` line 269 at r3](https://reviewable.io/reviews/dsccommunity/WSManDsc/106#-OGOlpVA4B56ubsTwVVH:-OGPpM60-prplHopZOsp:b-896fix) ([raw file](https://github.com/dsccommunity/WSManDsc/blob/514c668d3a651a2ab0c70d310b4acd6d9fe1c9dc/source/Classes/020.WSManListener.ps1#L269)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Suggest to use `Get-ComputerName` here instead (from DscResource.Common)
</blockquote></details>

Done.
___
*[`source/Classes/020.WSManListener.ps1` line 288 at r3](https://reviewable.io/reviews/dsccommunity/WSManDsc/106#-OGOmH9bBU8rXt1ozZVW:-OGPq8Bu-6l9JKpnf04R:bqvxyx9) ([raw file](https://github.com/dsccommunity/WSManDsc/blob/514c668d3a651a2ab0c70d310b4acd6d9fe1c9dc/source/Classes/020.WSManListener.ps1#L288)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

Do we need to validate here as well, that the required properties to remove an instance is actually set? :thinking:
</blockquote></details>

Same as the port above. Using `DscResource.Base` calling `Set()` calls `Get()` which calls `GetCurrentState()` which resolves those properties if not set or supplied. 

I also specifically did not touch the integration tests to remove the moving parts. 
___
*[`source/Private/Find-Certificate.ps1` line 24 at r3](https://reviewable.io/reviews/dsccommunity/WSManDsc/106#-OGOeUqC-UnodX1NNem0:-OGPrkhm9252iplzOBj4:bhfnff2) ([raw file](https://github.com/dsccommunity/WSManDsc/blob/514c668d3a651a2ab0c70d310b4acd6d9fe1c9dc/source/Private/Find-Certificate.ps1#L24)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

We also have a `Find-Certificate` in [CertificateDsc](https://github.com/dsccommunity/CertificateDsc/blob/0dcd82e1d6fb67e1f0f1dd597c8cbe3b97abb163/source/Modules/CertificateDsc.Common/CertificateDsc.Common.psm1#L280C10-L281C2) - it might be possible to concat these inte one public function in DscResource.Common. :thinking:  But out of scope for this PR.
</blockquote></details>

I'll create issues in the three repos for these.


<!-- Sent from Reviewable.io -->

dan-hughes commented:
@johlju, this is ready for another review.

I reverted back to using an Enum on one of the DSC properties. It cannot be nullable, but if the Enum values begin at 1, when the class is initialized without a value it will be 0.

This really only impacts optional properties and especially when there is an assert check. I added some logic to test if any of the properties are Enums and value = 0 which means it was not set by the user and can be ignored.

It seems to work, I have a feeling there is something I've not thought about though...

johlju commented:
I will try to get on this next, as soon as I can. But there is a lot a regular work this week.

dan-hughes commented:
That's fine. There is actually something I wanted to discuss/propose around DscResource.Base to enable the use of Enums for optional properties. 
I'll create a discussion in the correct repo.

dan-hughes commented:
@johlju this is ready for review again. Uses new base class functionality and preview DocGenerator which has the class enhancements.

dan-hughes reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/dsccommunity/WSManDsc/106)* status: all files reviewed, 2 unresolved discussions (waiting on @johlju and @PlagueHO)

___
*[`source/Classes/020.WSManListener.ps1` line 249 at r3](https://reviewable.io/reviews/dsccommunity/WSManDsc/106#-OGOl1QU2UkHPr2gXvVG:-OJDzgMJ5b3xMRFzF2yc:b-zytgc) ([raw file](https://github.com/dsccommunity/WSManDsc/blob/514c668d3a651a2ab0c70d310b4acd6d9fe1c9dc/source/Classes/020.WSManListener.ps1#L249)):*
<details><summary><i>Previously, johlju (Johan Ljunggren) wrote…</i></summary><blockquote>

I don't think that Port should be set in GetCurrentState, then we will enforce the port even if it was not passes by the user? :thinking: GetCurrentState is not meant to set the instance properties, just return current state. It is also more difficult to understand that Get() affects Set().

If the default port must be used I think it should be set in the Set-function prior to creating the listener. :thinking:
</blockquote></details>

This will include Address too. It can be moved into NewInstance(), which is where it is used.


<!-- Sent from Reviewable.io -->

johlju reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/dsccommunity/WSManDsc/106)* status: all files reviewed, 2 unresolved discussions (waiting on @dan-hughes and @PlagueHO)

___
*[`source/Classes/020.WSManListener.ps1` line 249 at r3](https://reviewable.io/reviews/dsccommunity/WSManDsc/106#-OGOl1QU2UkHPr2gXvVG:-OJEtoHm6JsV-hbuMZTX:b-bnjz5b) ([raw file](https://github.com/dsccommunity/WSManDsc/blob/514c668d3a651a2ab0c70d310b4acd6d9fe1c9dc/source/Classes/020.WSManListener.ps1#L249)):*
<details><summary><i>Previously, dan-hughes (Daniel Hughes) wrote…</i></summary><blockquote>

This will include Address too. It can be moved into NewInstance(), which is where it is used.
</blockquote></details>

Yes, same with Address, missed that. NewInsatnce() sounds good.


<!-- Sent from Reviewable.io -->

johlju commented:
/azp run

dan-hughes commented:
Not that I know of. I will raise as a separate PR if needs be.
